### PR TITLE
feat(resource_git_repository): add support for importing private repos

### LINKED ISSUE #180: Support Other Git service connections
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* azuredevops_XXXXX

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

### LINKED ISSUE #236: Support import private repository
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and Azure DevOps Provider) Version

Terraform v0.13.5
azurerm v2.36.0
random v3.0.0
azuredevops v0.1.0

### Affected Resource(s)

* `azuredevops_git_repository`

### Terraform Configuration Files

```hcl

provider "azurerm" {
  version = "~> 2.36.0"
  features {}
}

provider "azuredevops" {
  version = "= 0.1.0"
  org_service_url       = var.org_service_url 
  personal_access_token = var.personal_access_token
}

resource "azuredevops_project" "appproj" {
  name               = local.project_name
  description        = local.project_description
  visibility         = "private"
  version_control    = "Git"
  work_item_template = local.project_process
  features = {
      "testplans"    = "disabled"
      "artifacts"    = "disabled"
      "boards"       = "enabled" 
      "pipelines"    = "enabled"
      "repositories" = "enabled"
  }
}

## Add Project Owners to Project Administrator role
resource "azuredevops_user_entitlement" "owners" {
  principal_name = each.key
  for_each       = var.project_owners
}

resource "azuredevops_group" "project_owners_group" {
  scope        = azuredevops_project.appproj.id
  display_name = "Project Administrators"
}

resource "azuredevops_group_membership" "project_owners_membership" {
  for_each = var.project_owners
  group    = azuredevops_group.project_owners_group.descriptor
  members = [
    azuredevops_user_entitlement.owners[each.key].descriptor
  ]
}

resource "azuredevops_user_entitlement" "admins" {
  principal_name = each.key
  for_each       = var.project_admins
}

resource "azuredevops_group" "project_team_group" {
  scope        = azuredevops_project.appproj.id
  display_name = "${local.project_name} Team"
}

resource "azuredevops_group_membership" "project_admins_membership" {
  for_each = var.project_admins
  group    = azuredevops_group.project_team_group.descriptor
  members = [
    azuredevops_user_entitlement.admins[each.key].descriptor
  ]
}

data "azurerm_key_vault" "key_vault" {
  name                = module.subscription.kv_name
  resource_group_name = module.subscription.rgname
}

data "azurerm_key_vault_secret" "secret_key_vault" {
  name         = "secretname"  
  key_vault_id = data.azurerm_key_vault.key_vault.id
}

resource "azuredevops_serviceendpoint_azurerm" "endpointazure" {
  project_id            = azuredevops_project.appproj.id
  service_endpoint_name = local.project_spn_name

  credentials {
    serviceprincipalid  = var.spn_objectid
    serviceprincipalkey = data.azurerm_key_vault_secret.secret_key_vault.value
  }
  
  azurerm_spn_tenantid      = local.tenant_id
  azurerm_subscription_id   = module.subscription.subscription_guid
  azurerm_subscription_name = module.subscription.current_subscription_display_name
}

resource "azuredevops_git_repository" "repo" {
  project_id = azuredevops_project.appproj.id
  name       = "abcde"
  initialization {
    init_type = "Import"
    source_type = "Git"
    source_url = "<here comes a url of  a github repo in the same azure devops organisation, but in another project>"
  }
}

```

### Debug Output

 ```
2020/11/25 15:00:36 [DEBUG] azuredevops_git_repository.repo: applying the planned Create change
2020-11-25T15:00:36.195Z [DEBUG] plugin.terraform-provider-azuredevops_v0.1.0: 2020/11/25 15:00:36 [DEBUG] expandGitRepository: ID is empty (not set) 
2020/11/25 15:00:36 [DEBUG] azuredevops_git_repository.repo: apply errored, but we're indicating that via the Error pointer rather than returning it: Error import repository in Azure DevOps: <nil> 

```

### Panic Output

```
Error: Error import repository in Azure DevOps: <nil> 

  on main.tf line 138, in resource "azuredevops_git_repository" "repo":
 138: resource "azuredevops_git_repository" "repo" {
```

### Expected Behavior

The apply proceeds successfully, the repository is created and there is no error.

### Actual Behavior

What actually happens: 
1) terraform apply exits with the above mentioned error
2) the needed resource actually gets created and the repo is imported - can be seen in the Azure DevOps porta;
3) what is not expected - another empty repo is created with the same name as the project name. 

### Steps to Reproduce

1. `terraform apply`

### PR DESCRIPTION
**Pull request by dmccaffery**
feat(resource_git_repository): add support for importing private repos

* add support for specifying a `service_connection_id` for importing
  private repositories using the `serviceendpoint_generic_git` resource
* add acceptance test that imports a repository from azure devops via
  the `serviceendpoint_generic_git` resource
* fix an issue where errors in the import / initialization process could
  leave behind the uninitialized repository without being recorded in
  state
    
Closes: #236
Supersedes: #181

feat: add generic git service endpoint resource

* add `serviceendpoint_generic_git` resource to enable creation of a
  completely authentication service endpoint for any git repository url
* add acceptance tests for the resource
* add documentation for the resource

feat: add generic service endpoint resource

* add `serviceendpoint_generic` resource to enable creation of a
  completely generic authentication service endpoint for any server
* add acceptance tests for the resource
* add documentation for the resource

fix: add missing build tags on acceptance tests

* add build tags on `resource_repositorypolicy_author_email_patterns_test`
* add build tags on `resource_serviceendpoint_azuredevops_test`
* add build tags on `resource_serviceendpoint_ssh_test`

## All Submissions:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [X] All new and existing tests passed.
* [X] My code follows the code style of this project.
* [X] I ran lint checks locally prior to submission.
* [X] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Issue Number: #236 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [X] No

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

File: website/docs/r/git_repository.html.markdown
Comment (dmccaffery) [inline website/docs/r/git_repository.html.markdown:87]: this is a required field and doesn't actually default to Unitialized, so I removed it.

File: azuredevops/internal/service/git/resource_git_repository.go
Comment (dmccaffery) [inline azuredevops/internal/service/git/resource_git_repository.go:183]: Not sure how maintainers will feel about this, so I wanted to point it out. The benefit here is that terraform will taint the resource if an error is returned, which ensures that manual intervention is never required in the event of a failure. Any subsequent apply will recreate the resource. Prior to implementing this, I tried using a source url that had embedded basic auth username:password@, which failed to import and the repository was left behind requiring manual cleanup in order to run terraform again -- this resolves that issue.

**Comment by xuzhang3** [comment]
Hi @dmccaffery  Thanks for PR. 
The import authorization is based on service connection [Generic service](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#sep-generic). 
I would  suggest replace the username/password with service endpoint ID as the service request parameter asked
- Git repository resource can be decoupled from service endpoint resource
- Service endpoint can be reused in other resource.
- The new created service connection need to be deleted after import finished which will made the Git Repository resource more complex.   

There is a PR #181  working on generic git service endpoint, but no update for a long time. @paulbehrisch any update on PR #181 ?

**Comment by dmccaffery** [comment]
> Hi @dmccaffery Thanks for PR.
> The import authorization is based on service connection [Generic service](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#sep-generic).
> I would suggest replace the username/password with service endpoint ID as the service request parameter asked
> 
> * Git repository resource can be decoupled from service endpoint resource
> * Service endpoint can be reused in other resource.
> * The new created service connection need to be deleted after import finished which will made the Git Repository resource more complex.
> 
> There is a PR #181 working on generic git service endpoint, but no update for a long time. @paulbehrisch any update on PR #181 ?

I'm happy to implement this on top of an existing service connection once it is available. Re: deleting the service connection afterwards, there is a flag sent to the import request that causes the server to delete the service connection automatically after the import completes, so this doesn't need to be handled by us in terraform (other than setting the flag). We, of course, would not want to set this flag if passing in an existing service connection id instead of creating one within this resource.

It should be noted that I created this PR by reverse engineering the network traffic performed when doing an import via the UI, which always creates a new randomly named service connection and then passes in `DeleteServiceEndpointAfterImportIsDone` flag:

https://github.com/microsoft/terraform-provider-azuredevops/pull/402/files#diff-1c6ad48d90ac6ce910958242b9bcc943c4baea679f3028c62110654a227a01dbR217

We can put this PR on hold until the generic service connection is available. I can also look at taking over #181 if the original author and/or maintainers agree.

**Comment by dmccaffery** [comment]
@xuzhang3: I took the liberty to reimplementing this on top of a new `serviceendpoint_generic_git` resource as described in #118. I also added acceptance tests for everything and verified that an import of a private repository via the service connection works as expected in a real-world scenario.

Let me know if there is anything else I can do!

File: .azdo/ci.yml
Comment (dmccaffery) [inline .azdo/ci.yml:6]: eh... just noticed that my yaml formatter fired here -- let me know if I should revert.

File: website/docs/r/git_repository.html.markdown
Comment (dmccaffery) [inline website/docs/r/git_repository.html.markdown]: the username and password is incorrect; should be `service_connection_id`: will wait for other feedback before fixing to avoid multiple builds

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic.go
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic.go]: Password can also set to `Token Key`, the description should change to:
```
The password/token key to use for the generic service connection
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go]: Password can also set to `Token Key`, the description should change to:
```
The password/token key to use for the generic service connection
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go:1]: Can we add a new configuration to set the pipeline access flag?

image

File: website/docs/r/git_repository.html.markdown
Comment (xuzhang3) [reply website/docs/r/git_repository.html.markdown]: The `username/password` should be replaced  by `service_connection_id`.

File: website/docs/r/git_repository.html.markdown
Comment (xuzhang3) [inline website/docs/r/git_repository.html.markdown]: Can add  `$` back? Currently every document have `$`, we can remove it all at once in a new PR.

File: website/docs/r/serviceendpoint_generic.html.markdown
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_generic.html.markdown]:
```suggestion
  service_endpoint_name = "Sample Generic"
```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@dmccaffery Thanks for your PR, a few changes requested.
1. `azuredevops_serviceendpoint_generic` and `azuredevops_serviceendpoint_generic_git` document should indicate that user can use username/password or username/personal access token to create the service. Especially for GitHub user with 2-factor-auth enabled, they should create the service connection with `username/personal access token` not `username/password`, otherwise the authorization will failed.
2. `azuredevops_git_repository` can remove the `username/password` now,  the import authorization can use the service connection.

File: website/docs/r/git_repository.html.markdown
Comment (dmccaffery) [reply website/docs/r/git_repository.html.markdown]: yes; this was an accident (my markdown formatter removed it); apologies!

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go
Comment (dmccaffery) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_generic_git.go:1]: done; also updated tests

**Comment by dmccaffery** [comment]
@xuzhang3 all requested changes have been made; also rebased on top of master.

File: website/docs/r/serviceendpoint_generic_git.html.markdown
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_generic_git.html.markdown:43]: `enable_pipelines_access` not documented

Comment (dmccaffery) [reply website/docs/r/serviceendpoint_generic_git.html.markdown:43]: :feelsgood:

Comment (dmccaffery) [reply website/docs/r/serviceendpoint_generic_git.html.markdown:43]: fixed, and rebased

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
`serviceendpoint_generic.html.markdown` and `serviceendpoint_generic_git.html.markdown` document link need to be add to `azuredevops.erb`

**Comment by xuzhang3** [comment]
@dmccaffery I  have add the document link `serviceendpoint_generic.html.markdown` and `serviceendpoint_generic_git.html.markdown` document link to `azuredevops.erb`

**Comment by dmccaffery** [comment]
> @dmccaffery I  have add the document link `serviceendpoint_generic.html.markdown` and `serviceendpoint_generic_git.html.markdown` document link to `azuredevops.erb`

Thank you! I've been super busy (moving home) and haven't had a chance to circle back on this. Much appreciated!

**Comment by xuzhang3** [comment]
Acceptance test result:    
 
Git Repository
```log
==> Checking that code complies with gofmt requirements...
==> Sourcing .env file if avaliable
if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
        TF_ACC=1 go test -tags "all" $(go list ./azuredevops/internal/acceptancetests |grep -v 'vendor') -v -run=TestAccGitRepo -timeout 120m
=== RUN   TestAccGitRepository_DataSource
=== PAUSE TestAccGitRepository_DataSource
=== RUN   TestAccGitRepo_CreateAndUpdate
=== PAUSE TestAccGitRepo_CreateAndUpdate
=== RUN   TestAccGitRepo_Create_IncorrectInitialization
=== PAUSE TestAccGitRepo_Create_IncorrectInitialization
=== RUN   TestAccGitRepo_Create_Import
=== PAUSE TestAccGitRepo_Create_Import
=== RUN   TestAccGitRepo_RepoInitialization_Clean
=== PAUSE TestAccGitRepo_RepoInitialization_Clean
=== RUN   TestAccGitRepo_RepoInitialization_Uninitialized
=== PAUSE TestAccGitRepo_RepoInitialization_Uninitialized
=== RUN   TestAccGitRepo_RepoFork_BranchNotEmpty
=== PAUSE TestAccGitRepo_RepoFork_BranchNotEmpty
=== RUN   TestAccGitRepo_PrivateImport_BranchNotEmpty
=== PAUSE TestAccGitRepo_PrivateImport_BranchNotEmpty
=== CONT  TestAccGitRepository_DataSource
=== CONT  TestAccGitRepo_RepoInitialization_Uninitialized
=== CONT  TestAccGitRepo_Create_Import
=== CONT  TestAccGitRepo_PrivateImport_BranchNotEmpty
=== CONT  TestAccGitRepo_RepoInitialization_Clean
=== CONT  TestAccGitRepo_Create_IncorrectInitialization
=== CONT  TestAccGitRepo_RepoFork_BranchNotEmpty
=== CONT  TestAccGitRepo_CreateAndUpdate
--- PASS: TestAccGitRepo_Create_IncorrectInitialization (0.14s)
--- PASS: TestAccGitRepo_CreateAndUpdate (80.69s)
--- PASS: TestAccGitRepo_Create_Import (112.92s)
--- PASS: TestAccGitRepo_RepoInitialization_Clean (123.01s)
--- PASS: TestAccGitRepository_DataSource (184.16s)
--- PASS: TestAccGitRepo_RepoFork_BranchNotEmpty (204.82s)
--- PASS: TestAccGitRepo_RepoInitialization_Uninitialized (219.02s)
--- PASS: TestAccGitRepo_PrivateImport_BranchNotEmpty (236.31s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        237.321s
```

Generic service connection
```log
==> Checking that code complies with gofmt requirements...
==> Sourcing .env file if avaliable
if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
        TF_ACC=1 go test -tags "all" $(go list ./azuredevops/internal/acceptancetests |grep -v 'vendor') -v -run=TestAccServiceEndpointGeneric -timeout 120m
=== RUN   TestAccServiceEndpointGenericGit_basic
=== PAUSE TestAccServiceEndpointGenericGit_basic
=== RUN   TestAccServiceEndpointGenericGit_complete
=== PAUSE TestAccServiceEndpointGenericGit_complete
=== RUN   TestAccServiceEndpointGenericGit_update
=== PAUSE TestAccServiceEndpointGenericGit_update
=== RUN   TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== PAUSE TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== RUN   TestAccServiceEndpointGeneric_basic
=== PAUSE TestAccServiceEndpointGeneric_basic
=== RUN   TestAccServiceEndpointGeneric_complete
=== PAUSE TestAccServiceEndpointGeneric_complete
=== RUN   TestAccServiceEndpointGeneric_update
=== PAUSE TestAccServiceEndpointGeneric_update
=== RUN   TestAccServiceEndpointGeneric_RequiresImportErrorStep
=== PAUSE TestAccServiceEndpointGeneric_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointGenericGit_basic
=== CONT  TestAccServiceEndpointGeneric_complete
=== CONT  TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointGeneric_basic
=== CONT  TestAccServiceEndpointGenericGit_update
=== CONT  TestAccServiceEndpointGenericGit_complete
=== CONT  TestAccServiceEndpointGeneric_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointGeneric_update
--- PASS: TestAccServiceEndpointGeneric_basic (122.44s)
--- PASS: TestAccServiceEndpointGenericGit_complete (145.90s)
--- PASS: TestAccServiceEndpointGeneric_update (172.42s)
--- PASS: TestAccServiceEndpointGeneric_complete (175.42s)
--- PASS: TestAccServiceEndpointGenericGit_RequiresImportErrorStep (184.66s)
--- PASS: TestAccServiceEndpointGenericGit_basic (193.94s)
--- PASS: TestAccServiceEndpointGeneric_RequiresImportErrorStep (199.21s)
--- PASS: TestAccServiceEndpointGenericGit_update (206.17s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        207.033s
```

Generic Git service connection
```log
==> Checking that code complies with gofmt requirements...
==> Sourcing .env file if avaliable
if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
        TF_ACC=1 go test -tags "all" $(go list ./azuredevops/internal/acceptancetests |grep -v 'vendor') -v -run=TestAccServiceEndpointGenericGit -timeout 120m
=== RUN   TestAccServiceEndpointGenericGit_basic
=== PAUSE TestAccServiceEndpointGenericGit_basic
=== RUN   TestAccServiceEndpointGenericGit_complete
=== PAUSE TestAccServiceEndpointGenericGit_complete
=== RUN   TestAccServiceEndpointGenericGit_update
=== PAUSE TestAccServiceEndpointGenericGit_update
=== RUN   TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== PAUSE TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointGenericGit_basic
=== CONT  TestAccServiceEndpointGenericGit_RequiresImportErrorStep
=== CONT  TestAccServiceEndpointGenericGit_complete
=== CONT  TestAccServiceEndpointGenericGit_update
--- PASS: TestAccServiceEndpointGenericGit_complete (82.92s)
--- PASS: TestAccServiceEndpointGenericGit_RequiresImportErrorStep (94.03s)
--- PASS: TestAccServiceEndpointGenericGit_basic (147.33s)
--- PASS: TestAccServiceEndpointGenericGit_update (172.29s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        173.143s
```

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**
